src/commands/edge_app.rs

```diff
@@ -0,0 +1,241 @@
+
+use log::debug;
+use reqwest::header::HeaderMap;
+use reqwest::StatusCode;
+use serde::{Deserialize, Serialize};
+use serde_yaml;
+use std::fs;
+use std::fs::File;
+use std::io::Write;
+use std::path::Path;
+
+#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
+pub struct EdgeAppManifest {
+    #[serde(default)]
+    pub id: String,
+    #[serde(default)]
+    pub created_at: String,
+    #[serde(default)]
+    pub created_by: String,
+    #[serde(default)]
+    pub permissions: Vec<String>,
+    #[serde(default)]
+    pub updated_at: String,
+    pub name: String,
+    pub version: String,
+    pub description: String,
+    pub icon: String,
+    pub author: String,
+    pub homepage_url: String,
+}
+
+pub struct EdgeAppCommand {
+    authentication: Authentication,
+}
+
+impl EdgeAppCommand {
+    pub fn new(authentication: Authentication) -> Self {
+        Self { authentication }
+    }
+
+    pub fn init(self, path: &Path) -> Result<(), CommandError> {
+        let mut object = serde_yaml::to_value(EdgeAppManifest::default())?;
+        let map = object.as_mapping_mut().ok_or(CommandError::MissingField)?;
+
+        // following fields will be generated server side when publishing.
+        map.remove("id");
+        map.remove("created_at");
+        map.remove("created_by");
+        map.remove("updated_at");
+        map.remove("permissions");
+
+        let yaml = serde_yaml::to_string(map)?;
+        let input = File::create(path)?;
+        write!(&input, "{yaml}")?;
+
+        Ok(())
+    }
+
+    pub fn publish(self, path: &Path) -> Result<EdgeAppManifest, CommandError> {
+        let url = format!("{}/v4/edge_apps", &self.authentication.config.url);
+
+        let data = fs::read_to_string(path)?;
+        // by converting to struct we make sure there no extra fields and all required fields are present
+        let edge_app: EdgeAppManifest = serde_yaml::from_str(&data)?;
+        let mut binding = serde_json::to_value(edge_app)?;
+        let edge_app_json = binding.as_object_mut().ok_or(CommandError::Serde)?;
+
+        // Id can not be empty when posting. Depending on what we decide I should either raise an error
+        // or we allow users to supply an id.
+        if edge_app_json.contains_key("id")
+            && edge_app_json["id"]
+                .as_str()
+                .ok_or(CommandError::Serde)?
+                .is_empty()
+        {
+            edge_app_json.remove("id");
+        }
+
+        edge_app_json.remove("created_at");
+        edge_app_json.remove("created_by");
+        edge_app_json.remove("updated_at");
+        edge_app_json.remove("permissions");
+
+        let mut payload = HashMap::<String, String>::new();
+        // for now all values are fields in the manifest are required to be non-empty.
+        // if we change that we will need to have a list of required non-empty values.
+        for (k, v) in edge_app_json {
+            debug!("{k}{v:?}");
```

Comment:
Add at least some information on what's going on when debugging.

```suggestion
        // if we change that we will need to have a list of required non-empty values.
        debug!("Edge app JSON:");
        for (k, v) in edge_app_json {
            debug!("{k}{v:?}");
```

Reply:
debug does exactly that though, no? It will show context and what line it comes from